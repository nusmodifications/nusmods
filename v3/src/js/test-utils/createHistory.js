// @flow

import type { ContextRouter } from 'react-router-dom';
import _ from 'lodash';

// react-router-dom internal dependency, used here to construct the history
// object needed for testing. This is not added as a dev dependency to avoid
// version desync between the version depended on by react-router-dom
import createMemoryHistory from 'history/createMemoryHistory'; // eslint-disable-line import/no-extraneous-dependencies

type MatchShape = {
  params?: { [string]: ?string },
  isExact?: boolean,
};

export default function createHistory(
  initialEntries: string | string[] = '/',
  matchParams: MatchShape = {},
): ContextRouter {
  const entries = _.castArray(initialEntries);
  const history = createMemoryHistory({ initialEntries: entries });
  const { params = {}, isExact = true } = matchParams;

  const match = {
    params,
    isExact,
    path: entries[0],
    url: entries[0],
  };

  return {
    history,
    match,
    location: history.location,
  };
}
