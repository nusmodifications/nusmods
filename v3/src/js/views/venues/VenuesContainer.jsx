// @flow
import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import classnames from 'classnames';
import axios from 'axios';
import qs from 'query-string';
import Raven from 'raven-js';
import { pick } from 'lodash';

import type { MapStateToProps } from 'react-redux';
import type { ContextRouter } from 'react-router-dom';
import type { VenueInfo } from 'types/venues';
import type { Semester, Venue } from 'types/modules';

import ErrorPage from 'views/errors/ErrorPage';
import LoadingSpinner from 'views/components/LoadingSpinner';
import VenueList from 'views/venues/VenueList';
import SearchBox from 'views/components/SearchBox';

import config from 'config';
import nusmods from 'apis/nusmods';
import HistoryDebouncer from 'utils/HistoryDebouncer';

import styles from './VenuesContainer.scss';

type Props = {
  ...ContextRouter,
  activeSemester: Semester,
  urlVenue: ?Venue,
};

type State = {
  loading: boolean,
  venues: VenueInfo,
  error?: any,
  searchTerm: string,
  selectedVenue: Venue,
  selectedVenueElement?: HTMLElement,
};

const pageHead = (
  <Helmet>
    <title>Venues - {config.brandName}</title>
  </Helmet>
);

export class VenuesContainerComponent extends Component<Props, State> {
  history: HistoryDebouncer;

  constructor(props: Props) {
    super(props);

    const params = qs.parse(props.location.search);
    const selectedVenue = this.props.urlVenue || '';

    this.history = new HistoryDebouncer(props.history);
    this.state = {
      selectedVenue,
      loading: true,
      venues: {},
      searchTerm: params.q || selectedVenue,
    };
  }

  componentDidMount() {
    axios.get(nusmods.venuesUrl(this.props.activeSemester))
      .then(({ data }) => {
        this.setState({
          loading: false,
          venues: data,
        });
      })
      .catch((error) => {
        Raven.captureException(error);
        this.setState({ error });
      });
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.urlVenue) {
      this.setState({ selectedVenue: nextProps.urlVenue });
    }
  }

  componentDidUpdate() {
    if (this.state.selectedVenueElement) {
      // Scroll selected venue's row to just below the search box
      const scrollTargetY = this.state.selectedVenueElement.offsetTop;
      window.scrollTo(0, scrollTargetY);
    }
  }

  onVenueSelect = (selectedVenue: Venue, venueURL: string, selectedVenueElement: HTMLElement) => {
    this.setState({ selectedVenue, selectedVenueElement },
      () => this.updateURL(venueURL));
  };

  onSearch = (searchTerm: string) => {
    this.setState({ searchTerm }, () => this.updateURL());
  };

  updateURL(path?: string) {
    const { searchTerm } = this.state;
    const query = {};
    if (searchTerm) {
      query.q = searchTerm;
    }

    const pathname = path || this.props.location.pathname;

    this.history.push({
      ...this.props.location,
      search: qs.stringify(query),
      pathname,
    });
  }

  filteredVenues() {
    const { venues, searchTerm } = this.state;
    if (!venues) {
      return {};
    }

    if (searchTerm === '') {
      return venues;
    }

    const lowercaseSearchStr = searchTerm.toLowerCase();
    return pick(venues, Object.keys(venues).filter(name =>
      name.toLowerCase().includes(lowercaseSearchStr)));
  }

  render() {
    const { loading, error } = this.state;

    if (error) {
      return <ErrorPage error="cannot load venues info" eventId={Raven.lastEventId()} />;
    }

    if (loading) {
      return (
        <div>
          {pageHead}
          <LoadingSpinner />
        </div>
      );
    }

    const venues = this.filteredVenues();

    return (
      <div className={classnames('page-container', styles.pageContainer)}>
        {pageHead}

        <div className="row">
          <div className="col-sm-12">
            <SearchBox
              throttle={0}
              useInstantSearch
              initialSearchTerm={this.state.searchTerm}
              placeholder="Search for venues, e.g. LT27"
              onSearch={this.onSearch}
            />
            <VenueList
              venues={venues}
              expandedVenue={this.state.selectedVenue}
              onSelect={this.onVenueSelect}
            />
          </div>
        </div>
      </div>
    );
  }
}

export const mapStateToProps: MapStateToProps<*, *, *> = (state, ownProps) => {
  let venue: Venue = ownProps.match.params.venue;
  if (venue) {
    venue = decodeURIComponent(venue);
  }

  return {
    ...ownProps, // To silence a Flow error
    activeSemester: state.app.activeSemester,
    urlVenue: venue,
  };
};

export default connect(mapStateToProps)(withRouter(VenuesContainerComponent));
