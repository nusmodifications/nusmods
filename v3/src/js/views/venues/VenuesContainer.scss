@import "~styles/utils/modules-entry.scss";

.pageContainer {
  :global {
    animation-name: $page-entering-animation;
  }
}

.freeRoomIcon {
  vertical-align: middle;
}

.venueSearch {
  padding-right: ($grid-gutter-width / 2);

  h3 {
    margin-bottom: 0;
    font-weight: $font-weight-bold;
    font-size: $font-size-lg;
  }

  @include media-breakpoint-down(sm) {
    padding: 0.5rem ($grid-gutter-width / 2) 2.5rem;
  }
}

.availabilityToggle {
  margin-bottom: 0.5rem;
}

.searchBox {
  margin-bottom: 1rem;

  @include media-breakpoint-down(sm) {
    border-bottom: 1px solid var(--gray-light);
  }
}
