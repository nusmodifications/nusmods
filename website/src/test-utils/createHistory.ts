import type { RouteComponentProps, match as Match } from 'react-router-dom';
// react-router-dom internal dependency, used here to construct the history
// object needed for testing. This is not added as a dev dependency to avoid
// version desync between the version depended on by react-router-dom
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import _ from 'lodash';

type MatchShape = {
  params?: { [key: string]: string | null | undefined };
  isExact?: boolean;
};

// eslint-disable-next-line @typescript-eslint/ban-types
export default function createHistory<T = {}>(
  initialEntries: string | string[] = '/',
  matchParams: MatchShape = {},
): RouteComponentProps<T> {
  const entries = _.castArray(initialEntries);
  const history = createMemoryHistory({ initialEntries: entries });
  const { params = {}, isExact = true } = matchParams;

  const match: Match<T> = {
    // Not strictly type safe, but it's good enough for tests
    params: (params as unknown) as T,
    isExact,
    path: entries[0],
    url: entries[0],
  };

  return {
    history,
    match,
    location: history.location,
  };
}
