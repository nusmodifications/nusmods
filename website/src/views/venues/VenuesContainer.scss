@import '~styles/utils/modules-entry.scss';

$gutter: $grid-gutter-width / 2;
$sticky-top: $navbar-height + 1rem;
$venue-list-width: 16rem;
$venue-list-width-xl: 24rem;
$venue-list-width-xxl: 32rem;

// Layout for venues page is
// - On md and above - Outlook style two panel UI with the venue list on the left
//   and venue detail on the right
// - On sm and below - Single column venue list, with the venue detail appearing
//   in a modal when venue is selected
.pageContainer {
  :global {
    animation-name: $page-entering-animation;
  }
}

.venueDetail {
  &.mapExpanded {
    // Venue detail need to appear above everything else, including
    // page navigation when the map is expanded
    z-index: $venue-detail-expanded-z-index;
  }

  &:not(.mapExpanded) {
    // Only include the scrollable mixin when the map is not expanded
    // to prevent the map from being clipped in Safari Mobile
    @include scrollable-y;
  }
}

@include media-breakpoint-up(md) {
  .venuesList,
  .venueDetail {
    position: fixed;
    top: $sticky-top;
    bottom: 0;
    z-index: $venue-detail-z-index;
  }

  .venuesList {
    width: $venue-list-width;
    padding-bottom: 0; // For Microsoft Edge 18
    padding-bottom: env(safe-area-inset-bottom);
    border-right: 1px solid var(--gray-lighter);

    @include scrollable-y;
  }

  .venueDetail {
    right: 0; // For Microsoft Edge 18
    right: env(safe-area-inset-right);
    left: calc(#{$venue-list-width} + #{$gutter}); // For Microsoft Edge 18
    left: calc(env(safe-area-inset-left) + #{$venue-list-width} + #{$gutter});
    padding: 0 $gutter 0 $gutter; // For Microsoft Edge 18
    padding: 0 $gutter env(safe-area-inset-bottom) $gutter;
  }
}

@include media-breakpoint-up(xl) {
  .venuesList {
    width: $venue-list-width-xl;
  }

  .venueDetail {
    left: calc(#{$venue-list-width-xl} + #{$gutter}); // For Microsoft Edge 18
    left: calc(env(safe-area-inset-left) + #{$venue-list-width-xl} + #{$gutter});
  }
}

@media (min-width: 90rem) {
  .venuesList {
    width: $venue-list-width-xxl;
  }

  .venueDetail {
    left: calc(#{$venue-list-width-xxl} + #{$gutter}); // For Microsoft Edge 18
    left: calc(env(safe-area-inset-left) + #{$venue-list-width-xxl} + #{$gutter});
  }
}

.venueDetailModal {
  padding: 0;

  .closeButton {
    padding: 0.5rem 0;
    margin-bottom: 1rem;
    border-width: 0 0 1px;
    border-radius: 0;
  }
}

.venueSearch {
  margin: 0 0.5rem 1rem 0;

  h3 {
    margin-bottom: 0;
    font-weight: $font-weight-bold;
    font-size: $font-size-lg;
  }

  @include media-breakpoint-down(sm) {
    margin-right: 0;
  }
}

.searchBox {
  margin-bottom: 1rem;

  @include media-breakpoint-down(sm) {
    border-bottom: 1px solid var(--gray-light);
  }
}

.availabilitySearch {
  padding: ($border-radius + 0.2rem) 1rem 1rem;
  margin-top: -$border-radius;
  border: 1px solid theme-color();
  border-width: 0 1px 1px;
  border-radius: 0 0 $border-radius $border-radius;
}

.noVenueSelected {
  $icon-size: 6rem;

  opacity: 0.6;
  margin-top: 10vh;
  font-size: 1.4rem;
  text-align: center;

  svg {
    width: $icon-size;
    height: $icon-size;
    margin-bottom: 1rem;
  }
}
