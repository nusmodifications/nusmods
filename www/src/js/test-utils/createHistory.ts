// @flow

import type { ContextRouter, LocationShape } from 'react-router-dom';
import _ from 'lodash';

// react-router-dom internal dependency, used here to construct the history
// object needed for testing. This is not added as a dev dependency to avoid
// version desync between the version depended on by react-router-dom
import createMemoryHistory from 'history/createMemoryHistory'; // eslint-disable-line import/no-extraneous-dependencies

type MatchShape = {
  params?: { [string]: ?string },
  isExact?: boolean,
};

type HistoryEntry = string | LocationShape;

export default function createHistory(
  initialEntries: HistoryEntry | $ReadOnlyArray<HistoryEntry> = '/',
  matchParams: MatchShape = {},
): ContextRouter {
  const entries = _.castArray(initialEntries);
  const history = createMemoryHistory({ initialEntries: entries });
  const { params = {}, isExact = true } = matchParams;

  const match = {
    params,
    isExact,
    path: entries[0],
    url: entries[0],
  };

  return {
    history,
    match,
    location: history.location,
  };
}
