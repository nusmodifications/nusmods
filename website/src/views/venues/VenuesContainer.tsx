import {
  FC,
  unstable_useDeferredValue as useDeferredValue,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useHistory, useLocation, useParams } from 'react-router-dom';
import Loadable, { LoadingComponentProps } from 'react-loadable';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Location, locationsAreEqual } from 'history';
import classnames from 'classnames';
import axios from 'axios';
import qs from 'query-string';
import { isEqual, mapValues, pick, size } from 'lodash';

import type { TimePeriod, Venue, VenueDetailList, VenueSearchOptions } from 'types/venues';
import type { Subtract } from 'types/utils';

import deferComponentRender from 'views/hocs/deferComponentRender';
import ApiError from 'views/errors/ApiError';
import Warning from 'views/errors/Warning';
import LoadingOverlay from 'views/components/LoadingOverlay';
import LoadingSpinner from 'views/components/LoadingSpinner';
import SearchBox from 'views/components/SearchBox';
import { Clock } from 'react-feather';
import { venuePage } from 'views/routes/paths';
import Title from 'views/components/Title';

import config from 'config';
import nusmods from 'apis/nusmods';
import HistoryDebouncer from 'utils/HistoryDebouncer';
import { clampClassDuration, filterAvailability, searchVenue, sortVenues } from 'utils/venues';
import { convertIndexToTime } from 'utils/timify';

import AvailabilitySearch, { defaultSearchOptions } from './AvailabilitySearch';
import VenueDetailsPane from './VenueDetailsPane';
import VenueList from './VenueList';
import VenueLocation from './VenueLocation';
import styles from './VenuesContainer.scss';

type Params = {
  q: string;
  venue: string;
};

type LoadedProps = { venues: VenueDetailList };
type Props = LoadedProps;

export const VenuesContainerComponent: FC<Props> = ({ venues }) => {
  const history = useHistory();
  const debouncedHistory = useMemo(() => new HistoryDebouncer(history), [history]);
  const location = useLocation();
  const matchParams = useParams<Params>();

  const [searchQuery, setSearchQuery] = useState<string>(() => qs.parse(location.search).q || '');

  const [isAvailabilityEnabled, setIsAvailabilityEnabled] = useState(() => {
    const params = qs.parse(location.search);
    return !!(params.time && params.day && params.duration);
  });

  const [searchOptions, setSearchOptions] = useState(() => {
    const params = qs.parse(location.search);
    // Extract searchOptions from the query string if they are present
    return isAvailabilityEnabled
      ? (mapValues(pick(params, ['time', 'day', 'duration']), (i) =>
          parseInt(i, 10),
        ) as VenueSearchOptions)
      : defaultSearchOptions();
  });
  const [pristineSearchOptions, setPristineSearchOptions] = useState(() => !isAvailabilityEnabled);

  const deferredSearchQuery = useDeferredValue(searchQuery);
  const deferredIsAvailabilityEnabled = useDeferredValue(isAvailabilityEnabled);
  const deferredSearchOptions = useDeferredValue(searchOptions);
  const isPending =
    searchQuery !== deferredSearchQuery ||
    isAvailabilityEnabled !== deferredIsAvailabilityEnabled ||
    searchOptions !== deferredSearchOptions;

  // TODO: Check if this actually does anything useful
  useEffect(() => {
    VenueLocation.preload();
  }, []);

  const onFindFreeRoomsClicked = useCallback(() => {
    setIsAvailabilityEnabled(!isAvailabilityEnabled);
    if (pristineSearchOptions && !isAvailabilityEnabled) {
      // Only reset search options if the user has never changed it, and if the
      // search box is being opened. By resetting the option when the box is
      // opened, the time when the box is opened will be used, instead of the
      // time when the page is loaded.
      setSearchOptions(defaultSearchOptions());
    }
  }, [isAvailabilityEnabled, pristineSearchOptions]);

  const onAvailabilityUpdate = useCallback(
    (newSearchOptions: VenueSearchOptions) => {
      if (!isEqual(newSearchOptions, searchOptions)) {
        setSearchOptions(clampClassDuration(newSearchOptions));
        setPristineSearchOptions(false); // user changed searchOptions
      }
    },
    [searchOptions],
  );

  const highlightPeriod = useMemo<TimePeriod | undefined>(() => {
    if (!deferredIsAvailabilityEnabled) return undefined;

    return {
      day: deferredSearchOptions.day,
      startTime: convertIndexToTime(deferredSearchOptions.time * 2),
      endTime: convertIndexToTime(
        2 * (deferredSearchOptions.time + deferredSearchOptions.duration),
      ),
    };
  }, [
    deferredIsAvailabilityEnabled,
    deferredSearchOptions.day,
    deferredSearchOptions.duration,
    deferredSearchOptions.time,
  ]);

  const selectedVenue = useMemo<Venue | undefined>(
    () => (matchParams.venue ? decodeURIComponent(matchParams.venue) : undefined),
    [matchParams.venue],
  );

  // Sync URL with component state
  useEffect(() => {
    let query: Partial<Params> = {};
    if (deferredSearchQuery) query.q = deferredSearchQuery;
    if (deferredIsAvailabilityEnabled) query = { ...query, ...deferredSearchOptions };
    const search = qs.stringify(query);

    const pathname = venuePage(selectedVenue);

    const proposedLocation: Location = {
      ...history.location,
      search,
      pathname,
    };

    if (!locationsAreEqual(history.location, proposedLocation)) {
      // TODO: Consider replacing our debounced history with
      // `React.useTransition` or a React scheduler low priority callback.
      debouncedHistory.push(proposedLocation);
    }
  }, [
    debouncedHistory,
    deferredIsAvailabilityEnabled,
    deferredSearchOptions,
    deferredSearchQuery,
    history,
    selectedVenue,
  ]);

  const matchedVenues = useMemo(() => {
    const matched = searchVenue(venues, deferredSearchQuery);
    return deferredIsAvailabilityEnabled
      ? filterAvailability(matched, deferredSearchOptions)
      : matched;
  }, [deferredIsAvailabilityEnabled, deferredSearchOptions, deferredSearchQuery, venues]);
  const matchedVenueNames = useMemo(() => matchedVenues.map(([venue]) => venue), [matchedVenues]);

  function renderSearch() {
    return (
      <div className={styles.venueSearch}>
        <h3>Venue Search</h3>

        <SearchBox
          className={styles.searchBox}
          value={searchQuery}
          placeholder="e.g. LT27"
          onChange={setSearchQuery}
        />

        <button
          className={classnames(
            'btn btn-block btn-svg',
            isAvailabilityEnabled ? 'btn-primary' : 'btn-outline-primary',
          )}
          onClick={onFindFreeRoomsClicked}
          type="button"
        >
          <Clock className="svg" /> Find free rooms
        </button>

        {isAvailabilityEnabled && (
          <div className={styles.availabilitySearch}>
            <AvailabilitySearch
              isEnabled={isAvailabilityEnabled}
              searchOptions={searchOptions}
              onUpdate={onAvailabilityUpdate}
            />
          </div>
        )}
      </div>
    );
  }

  const disableAvailability = useCallback(() => setIsAvailabilityEnabled(false), []);
  function renderNoResult() {
    const unfilteredCount = size(matchedVenues);
    return (
      <>
        <Warning message="No matching venues found" />
        {!!unfilteredCount && isAvailabilityEnabled && (
          <p className="text-center text-muted">
            {unfilteredCount === 1
              ? 'There is a venue that is not shown because it is not free'
              : `There are ${unfilteredCount} venues that are not shown because they are not free`}
            <br />
            <button type="button" className="btn btn-link" onClick={disableAvailability}>
              Show all rooms
            </button>
          </p>
        )}
      </>
    );
  }

  return (
    <div className={classnames('page-container', styles.pageContainer)}>
      <Title>Venues</Title>

      <div className={styles.venuesList}>
        {renderSearch()}

        <div className="position-relative">
          {isPending && (
            <LoadingOverlay deferred>
              <LoadingSpinner />
            </LoadingOverlay>
          )}
          {!isPending && size(matchedVenues) === 0 && renderNoResult()}
          <VenueList venues={matchedVenueNames} selectedVenue={selectedVenue} />
        </div>
      </div>

      <VenueDetailsPane
        highlightPeriod={highlightPeriod}
        matchedVenues={matchedVenues}
        selectedVenue={selectedVenue}
        venues={venues}
      />
    </div>
  );
};

const AsyncVenuesContainer = Loadable.Map<Subtract<Props, LoadedProps>, LoadedProps>({
  loader: {
    venues: async () => {
      const response = await axios.get(nusmods.venuesUrl(config.semester));
      return sortVenues(response.data);
    },
  },
  loading: (props: LoadingComponentProps) => {
    if (props.error) {
      return <ApiError dataName="venue information" retry={props.retry} />;
    }

    if (props.pastDelay) {
      return <LoadingSpinner />;
    }

    return null;
  },
  render({ venues }, props) {
    return <VenuesContainerComponent venues={venues} {...props} />;
  },
});

export default deferComponentRender(AsyncVenuesContainer);

export function preload() {
  AsyncVenuesContainer.preload();
  VenueLocation.preload();
}
