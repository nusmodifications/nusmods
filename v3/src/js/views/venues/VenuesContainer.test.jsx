// @flow
import React from 'react';
import { shallow } from 'enzyme';
import createHistory from 'history/createMemoryHistory'; // eslint-disable-line import/no-extraneous-dependencies

import type { Venue } from 'types/modules';

import venueInfo from '__mocks__/venueInformation.json';
import { VenuesContainerComponent, mapStateToProps } from './VenuesContainer';

function createComponent(urlVenue: ?Venue) {
  const history = createHistory();
  const mockMatch = {
    path: '/',
    url: '/',
    isExact: true,
    params: {},
  };

  return shallow(
    <VenuesContainerComponent
      history={history}
      location={history.location}
      match={mockMatch}
      urlVenue={urlVenue}
      activeSemester={1}
    />,
  );
}

describe('VenuesContainer', () => {
  test('#filteredVenues() should filter venues correctly', () => {
    const component = createComponent().instance();
    component.setState({ venues: venueInfo });

    // No filter
    component.setState({ searchTerm: '' });
    expect(Object.keys(component.filteredVenues())).toMatchSnapshot();

    // Unique venue
    component.setState({ searchTerm: 'S11-0302' });
    expect(Object.keys(component.filteredVenues())).toEqual(['S11-0302']);

    // Unique venue with wrong case
    component.setState({ searchTerm: 's11-0302' });
    expect(Object.keys(component.filteredVenues())).toEqual(['S11-0302']);

    // Substring
    component.setState({ searchTerm: 'T1' });
    expect(Object.keys(component.filteredVenues())).toEqual(['LT17', 'LT1']);

    // Venue which does not exist
    component.setState({ searchTerm: 'covfefe' });
    expect(Object.keys(component.filteredVenues())).toEqual([]);
  });

  test('#onVenueSelect() should change the URL when a venue is clicked', () => {
    const component = createComponent().instance();
    component.onVenueSelect('LT17', '/venues/LT17');
    expect(component.state.selectedVenue).toEqual('LT17');
    expect(component.props.history.location.pathname).toEqual('/venues/LT17');
  });

  describe('URL handling', () => {
    test('it should only filter venues by venue in URL on init if present and no query string', () => {
      // Filter by URL venue on init and no query string
      const wrapper = createComponent('Interwebs');
      const instance = wrapper.instance();
      expect(instance.state.searchTerm).toEqual('Interwebs');

      // Don't filter by URL venue when venue clicked
      wrapper.setProps({
        urlVenue: 'covfefe',
      });
      expect(instance.state.searchTerm).toEqual('Interwebs');
    });

    test('it should select venue in URL if present and appropriate', () => {
      const component = createComponent('Interwebs');
      const instance = component.instance();

      // Select URL venue on init
      expect(instance.state.selectedVenue).toEqual('Interwebs');

      // Select URL venue when changed
      component.setProps({ urlVenue: 'covFEFE' });
      expect(instance.state.selectedVenue).toEqual('covFEFE');
    });

    test('it should neither select nor filter if no venue is present in URL', () => {
      const component = createComponent();
      const instance = component.instance();

      // No URL on init
      expect(instance.state.searchTerm).toEqual('');
      expect(instance.state.selectedVenue).toEqual('');

      // No URL on props set
      component.setProps({
        history: { action: 'POP' },
        urlVenue: undefined, // No react-router match
      });
      expect(instance.state.searchTerm).toEqual('');
      expect(instance.state.selectedVenue).toEqual('');
    });
  });

  describe('#updateURL()', () => {
    test('it should update URL query', () => {
      const wrapper = createComponent();
      const instance = wrapper.instance();

      instance.setState({ searchTerm: 'covfefe' });
      expect(instance.props.history.location.search).toBe('');

      // Should set query string
      instance.updateURL();
      wrapper.setProps({ location: instance.props.history.location });
      expect(instance.props.history.location.search).toBe('?q=covfefe');

      // Should decode special chars
      instance.setState({ searchTerm: 'Cdat/overThar1!' });
      instance.updateURL();
      wrapper.setProps({ location: instance.props.history.location });
      expect(instance.props.history.location.search).toBe('?q=Cdat%2FoverThar1%21');

      // Should clear query string
      instance.setState({ searchTerm: '' });
      instance.updateURL();
      wrapper.setProps({ location: instance.props.history.location });
      expect(instance.props.history.location.search).toBe('');
    });

    test('it should update URL path if present', () => {
      const wrapper = createComponent();
      const instance = wrapper.instance();

      // Clear current value
      instance.updateURL('/');
      wrapper.setProps({ location: instance.props.history.location });
      expect(instance.props.history.location.pathname).toBe('/');

      // Update value
      instance.updateURL('/venues/covfefe');
      wrapper.setProps({ location: instance.props.history.location });
      expect(instance.props.history.location.pathname).toBe('/venues/covfefe');

      // Should not change path if none is provided
      instance.updateURL();
      wrapper.setProps({ location: instance.props.history.location });
      expect(instance.props.history.location.pathname).toBe('/venues/covfefe');
    });
  });

  test('#mapStateToProps() should set semester and decode URL venue', () => {
    const state = { app: { activeSemester: 1 } };
    const ownProps = { match: { params: { venue: 'Cdat%2FoverThar1%21' } } };
    const mappedProps = mapStateToProps(state, ownProps);
    expect(mappedProps).toMatchObject(ownProps);
    // Should set activeSemester
    expect(mappedProps).toHaveProperty('activeSemester', 1);
    // Should decode urlVenue
    expect(mappedProps).toHaveProperty('urlVenue', 'Cdat/overThar1!');
  });
});
