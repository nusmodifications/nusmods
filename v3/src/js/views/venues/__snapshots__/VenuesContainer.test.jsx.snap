// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VenuesContainer #filteredVenues() should filter venues correctly 1`] = `
Array [
  "LT17",
  "lt2",
  "S11-0302",
  "LT1",
  "CQT/SR0622",
]
`;
